File: bots/intgrah/v54.7.7/builder/tasks/econ/chains/extend_chain_in_range.py
"""Lay a conveyor segment from `dangling_output` toward its sink, only
when the dangling end is within builder vision. The cached
`dangling_output` is refreshed every turn by `update_dangling` (no
stickiness), and the proximity gate inside the picker ensures only the
rightful claimant builder considers a given end.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, override

from builder.chain_routing import extend_chain
from builder.tasks.rejected import Reason, TaskRejectedError

if TYPE_CHECKING:
    from cambc import Controller, Position

    from builder import Builder


class TaskRejectedNoDanglingOutputError(TaskRejectedError):
    @override
    def reason(self) -> Reason:
        return "no dangling output"


class TaskRejectedOutOfVisionError(TaskRejectedError):
    def __init__(self, target: Position) -> None:
        self.target = target

    @override
    def reason(self) -> Reason:
        return "dangling {target} not in vision", {"target": self.target}


def extend_chain_in_range(self: Builder, ct: Controller) -> None:
    if self.dangling_output is None:
        raise TaskRejectedNoDanglingOutputError
    if not ct.is_in_vision(self.dangling_output):
        raise TaskRejectedOutOfVisionError(self.dangling_output)
    extend_chain(self, ct, self.dangling_output)

File: bots/rs/v56/src/builder/tasks/econ/chains/extend_chain_in_range.rs
//! Lay a conveyor segment from `dangling_output` toward its sink, only
//! when the dangling end is within builder vision. The cached
//! `dangling_output` is refreshed every turn by `update_dangling` (no
//! stickiness), and the proximity gate inside the picker ensures only the
//! rightful claimant builder considers a given end.

use cambc::{Controller, ControllerApi};

use crate::builder::Builder;
use crate::builder::chain_routing::extend_chain;
use crate::builder::tasks::rejected::{TaskRejected, TaskResult};

pub fn extend_chain_in_range(self_: &mut Builder, ct: &mut Controller<'_>) -> TaskResult {
    let Some(dangling) = self_.dangling_output else {
        return Some(TaskRejected::new("no dangling output"));
    };
    if !pyrust::unwrap!(ct.is_in_vision(dangling)) {
        return Some(TaskRejected::from_string(format!(
            "dangling {dangling:?} not in vision"
        )));
    }
    extend_chain(self_, ct, dangling)
}
